stan/math/rev/mat/functor/de_integrator.hpp

```diff
@@ -0,0 +1,149 @@
+/*
+ * The original code on which this file is based is from John Cook,
+ * which is licensed under the 2-clause BSD license, reproduced below:
+ *
+ * Copyright (c) 2015, John D. Cook
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ * Redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+*/
+#ifndef STAN_MATH_REV_MAT_FUNCTOR_DEINTEGRATOR_HPP
+#define STAN_MATH_REV_MAT_FUNCTOR_DEINTEGRATOR_HPP
+
+#include "de_integrator_constants.hpp"
+#include <cmath>
+#include <cfloat>
+
+namespace stan {
+
+  namespace math {
+
+    /**
+     * Double Exponential Integrator.

```

Comment:
Is there a citation for the algorithm being used?  

Are there conditions on the function `f`?  

Can the upper and lower limits be infiite?

Must absolute and relative error requirements both be met or is one enough?